plutus-core/executables/debugger/Draw.hs

```diff
@@ -0,0 +1,164 @@
+                    (BE.renderEditor (drawDocumentWithHighlight (st ^. dsUplcHighlight)))
+                    (st ^. dsUplcEditor)
+                , B.padTop (B.Pad 1) . BC.hCenter $ B.str cursorPos
+                ]
+    sourceEditor =
+        BB.borderWithLabel (B.txt "Source program") $
+            B.withFocusRing
+                focusRing
+                (BE.renderEditor (B.txt . Text.unlines))
+                (st ^. dsSourceEditor)
+    returnValueEditor =
+        BB.borderWithLabel (B.txt "UPLC value being returned") $
+            B.withFocusRing
+                focusRing
+                (BE.renderEditor (B.txt . Text.unlines))
+                (st ^. dsReturnValueEditor)
+    cekStateEditor =
+        BB.borderWithLabel (B.txt "CEK machine state") $
+            B.withFocusRing
+                focusRing
+                (BE.renderEditor (B.txt . Text.unlines))
+                (st ^. dsCekStateEditor)
+    ui =
+        B.vBox
+            [ BC.center uplcEditor B.<+> BC.center sourceEditor
+            , B.vLimit (st ^. dsVLimitBottomEditors) $
+                BC.center returnValueEditor B.<+> BC.center cekStateEditor
+            , footer
+            ]
+
+-- | Draw a document, a consecutive part of which may be highlighted.
+drawDocumentWithHighlight ::
+    forall n.
+    -- | The part of the document to be highlighted.
+    Maybe HighlightSpan ->
+    -- | The document to draw, one `Text` per line.
+    [Text] ->
+    B.Widget n
+drawDocumentWithHighlight = \case
+    Nothing -> B.txt . Text.unlines
+    Just (HighlightSpan (B.Location (sLine, sCol)) eLoc) ->
+        let toWidgets :: [Text] -> [B.Widget n]
+            toWidgets lns =
+                let alg :: (Text, Int) -> B.Widget n
+                    alg (line, lineNo)
+                        | lineNo < sLine || lineNo > eLine = B.txt line
+                        -- The current line (either entirely or part of it) needs to be highlighted.
+                        | otherwise =
+                            let s =
+                                    if lineNo > sLine
+                                        then 0
+                                        else sCol - 1
+                                e =
+                                    if lineNo < eLine
+                                        then Text.length line
+                                        else eCol
+                             in highlightLine line s e
+                      where
+                        B.Location (eLine, eCol) =
+                            fromMaybe
+                                (B.Location (sLine, Text.length line))
+                                eLoc
+                 in fmap alg (zip lns [1 ..])
+         in B.vBox
+                . toWidgets
+                -- This is needed for empty lines to be rendered correctly.
+                -- `Brick.Widgets.Core.txt` does the same via `fixEmpty`.
+                . fmap (\t -> if Text.null t then " " else t)
```

Comment:
I note that nothing in `toWidgets` touches the `Text`s, only `highlightLine` does. So maybe this workaround could go in `highlightLine`? Indeed, it's not clear to me which bit will go wrong, making the workaround more localised might help with that. Is it just that we need

```
txt' t | T.null t = txt " "
txt' t = txt t
```
?

Reply:
> txt' t | T.null t = txt " "
txt' t = txt t

Exactly. But we also don't want `txt'` everywhere. It is just when the text we are rendering has _multiple_ lines, then the empty lines need to be fixed. And moving the workaround to `highlightLine` will break this:
> | lineNo < sLine || lineNo > eLine = B.txt line